SparseGrids/tsgLinearSolvers.cpp

```diff
@@ -237,6 +256,97 @@ void TasmanianTridiagonalSolver::decompose(int n, std::vector<double> &d, std::v
     }
 }
 
+void TasmanianTridiagonalSolver::decompose2(std::vector<double> &diag, std::vector<double> &off_diag, const double mu0,
+                                            std::vector<double> &nodes, std::vector<double> &weights) {
+
+    // Ensure compatibility with the ALGOL implementation.
+    // NOTE: diag and off_diag are 1-indexed, while nodes and weights are 0-indexed after this step.
+    size_t n = diag.size();
+    if (off_diag.size() != n-1) {
+        throw std::invalid_argument("ERROR: off_diag.size() must be equal to diag.size()-1");
```

Comment:
Use an `assert()`  in the internal API. Raising complex exceptions and doing the corresponding checks is wasteful, `assert()` is a cheap way for us to keep out code sane and the user can remove them at runtime. After all, if the rest of the code is correct, this will never happer.

* For external API, use informative error checking since the can mess up the inputs in all sorts of ways.
* For internal API, use `assert()` and only if you really need to use it.

Reply:
Changed. Note that I had to add `cassert` to the list of includes.